### PR TITLE
[Single File] Allow loading T5 encoder in mixed precision 

### PR DESCRIPTION
**Pull request by DN6**
# What does this PR do?
When loading the T5 encoder in SD3 with `from_pretrained` certain layers are are kept in fp32. Currently this is not the case with single file loading as all modules are cast to `torch_dtype` when loading the pipeline. This is leading to NaNs being produced by the T5 encoder when loading from single file, leading to a black image. 

This PR 

1. Adds a check to T5 loading to convert the required modules to fp32
2. Removes casting the entire pipeline to `torch_dtype`. This is now only handled at the model level. 
2. Adds additional tests to the single file suite to check that dtypes are set correctly in the pipeline   

Single file tests for SD3 need to be added. I'll handle that in a follow up PR. 

Fixes # (issue)
https://github.com/huggingface/diffusers/issues/8759

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [ ] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/diffusers/loaders/single_file.py [hunk 1]
```diff
@@ -555,7 +555,4 @@ def load_module(name, value):
 
         pipe = pipeline_class(**init_kwargs)
 
-        if torch_dtype is not None:
-            pipe.to(dtype=torch_dtype)
```
Comment (sayakpaul) [inline src/diffusers/loaders/single_file.py:559]: How is this handled then?

Comment (DN6) [reply src/diffusers/loaders/single_file.py:559]: At the model level by passing in `dtype` to their respective loading methods

File: src/diffusers/loaders/single_file_utils.py [hunk 2]
```diff
@@ -1808,4 +1808,17 @@ def create_diffusers_t5_model_from_checkpoint(
 
     else:
         model.load_state_dict(diffusers_format_checkpoint)
+
```
Comment (yiyixuxu) [inline src/diffusers/loaders/single_file_utils.py:1811]: ohh is this related to this? 
https://github.com/huggingface/diffusers/issues/8604#issuecomment-2174675284

**Review by yiyixuxu — APPROVED** [review]
thanks!

File: src/diffusers/loaders/single_file_utils.py (hunk 2 above)
Comment (DN6) [reply src/diffusers/loaders/single_file_utils.py:1811]: Yeah exactly. When you cast the entire model to `fp16` the images will be corrupted. You have to use the logic in `from_pretrained`.
